src/main/java/picard/sam/AbstractAlignmentMerger.java

```diff
@@ -751,54 +765,103 @@ private void transferAlignmentInfoToPairedRead(final SAMRecord firstUnaligned, f
-                    final int posDiff = pos.getAlignmentEnd() - neg.getAlignmentEnd();
-                    final int negDiff = pos.getAlignmentStart() - neg.getAlignmentStart();
-
-                    if (posDiff > 0) {
-                        final List<CigarElement> elems = new ArrayList<>(pos.getCigar().getCigarElements());
-                        Collections.reverse(elems);
-                        final int clipped = lengthOfSoftClipping(elems.iterator());
-                        final int clipFrom = pos.getReadLength() - posDiff - clipped + 1;
-                        CigarUtil.softClip3PrimeEndOfRead(pos, Math.min(pos.getReadLength(), clipFrom));
-                    }
+                    final int posClipFrom = getReadPositionAtReferencePositionIgnoreSoftClips(pos, neg.getUnclippedEnd() + 1);
+                    int negClipFrom = getReadPositionAtReferencePositionIgnoreSoftClips(neg, pos.getUnclippedStart() - 1);
+                    negClipFrom = negClipFrom > 0 ? (neg.getReadLength() + 1) - negClipFrom : 0;
 
-                    if (negDiff > 0) {
-                        final int clipped = lengthOfSoftClipping(neg.getCigar().getCigarElements().iterator());
-                        final int clipFrom = neg.getReadLength() - negDiff - clipped + 1;
-                        CigarUtil.softClip3PrimeEndOfRead(neg, Math.min(neg.getReadLength(), clipFrom));
+                    if(posClipFrom > 0) {
+                        clipRead(pos, posClipFrom, useHardClipping);
+                    }
+                    if(negClipFrom > 0) {
+                        clipRead(neg, negClipFrom, useHardClipping);
                     }
                 }
             }
         }
     }
 
-    /** Returns the number of soft-clipped bases until a non-soft-clipping element is encountered. */
-    private static int lengthOfSoftClipping(Iterator<CigarElement> iterator) {
-        int clipped = 0;
-        while (iterator.hasNext()) {
-            final CigarElement elem = iterator.next();
-            if (elem.getOperator() != CigarOperator.SOFT_CLIP && elem.getOperator() != CigarOperator.HARD_CLIP) {
-                break;
-            }
-            if (elem.getOperator() == CigarOperator.SOFT_CLIP) {
-                clipped = elem.getLength();
+    private static int getReadPositionAtReferencePositionIgnoreSoftClips(final SAMRecord rec, final int pos) {
+        final int readPosition;
+        final Cigar oldCigar = rec.getCigar();
+        final int oldStart = rec.getAlignmentStart();
+        final Cigar newCigar = new Cigar();
+        List<CigarElement> cigarElements = new ArrayList<>(oldCigar.getCigarElements()); //need to be modifiable
+        int posShift = 0;
+        boolean foundNonClip = false;
+        for (final CigarElement cigarElement : cigarElements) {
+            final CigarOperator op = cigarElement.getOperator();
+
+            if (op == CigarOperator.SOFT_CLIP) {
```

Comment:
I think some comments about what's happening here would be appreciated by a future developer...